In [1]:
import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow import keras
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.model_selection import train_test_split

In [2]:
pd.set_option('display.max_colwidth', 9999999)

# View Data

### Training Movie Dataset

In [3]:
df_training_items = pd.read_csv('./data/movies_for_training.csv')
df_training_items.head(5)

movie id  year  ave rating  Action  Adventure  Animation  Children  Comedy  \
0      6874  2003    3.961832       1          0          0         0       0   
1      8798  2004    3.761364       1          0          0         0       0   
2     46970  2006    3.250000       1          0          0         0       1   
3     48516  2006    4.252336       0          0          0         0       0   
4     58559  2008    4.238255       1          0          0         0       0   

   Crime  Documentary  Drama  Fantasy  Horror  Mystery  Romance  Sci-Fi  \
0      1            0      0        0       0        0        0       0   
1      1            0      1        0       0        0        0       0   
2      0            0      0        0       0        0        0       0   
3      1            0      1        0       0        0        0       0   
4      1            0      1        0       0        0        0       0   

   Thriller  
0         1  
1         1  
2         0  
3         1  
4         0

In [4]:
df_training_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50884 entries, 0 to 50883
Data columns (total 17 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   movie id     50884 non-null  int64  
 1   year         50884 non-null  int64  
 2   ave rating   50884 non-null  float64
 3   Action       50884 non-null  int64  
 4   Adventure    50884 non-null  int64  
 5   Animation    50884 non-null  int64  
 6   Children     50884 non-null  int64  
 7   Comedy       50884 non-null  int64  
 8   Crime        50884 non-null  int64  
 9   Documentary  50884 non-null  int64  
 10  Drama        50884 non-null  int64  
 11  Fantasy      50884 non-null  int64  
 12  Horror       50884 non-null  int64  
 13  Mystery      50884 non-null  int64  
 14  Romance      50884 non-null  int64  
 15  Sci-Fi       50884 non-null  int64  
 16  Thriller     50884 non-null  int64  
dtypes: float64(1), int64(16)
memory usage: 6.6 MB


In [5]:
df_training_items['movie id'].value_counts()

movie id
5669     1160
8464     1000
8622      740
4306      680
79132     572
         ... 
69784      10
62155      10
32029      10
55830      10
68793      10
Name: count, Length: 847, dtype: int64

**There are 50884 rows to match with the viewer dataset and the rating dataset but only 847 movies**

### Training Viewer Dataset

In [6]:
df_training_users = pd.read_csv('./data/viewers_for_training.csv')
df_training_users.sample(5)

user id  rating count  rating ave  Action  Adventure  Animation  \
12414      169            80        4.15    3.68       4.18       4.75   
7937       105           320        3.90    3.83       3.82       3.85   
41483      510            15        2.73    2.80       2.25       0.00   
29113      365           157        2.78    2.98       2.80       2.36   
41249      509           239        3.18    3.17       3.30       3.50   

       Children  Comedy  Crime  Documentary  Drama  Fantasy  Horror  Mystery  \
12414      4.75    4.24   3.83         4.75   4.30     4.58    4.00     4.17   
7937       3.86    3.76   4.00         4.12   4.02     3.75    3.97     4.19   
41483      0.00    3.00   3.00         0.00   2.83     2.75    1.50     2.50   
29113      2.53    3.04   2.91         1.25   2.50     2.91    2.56     2.25   
41249      3.31    3.08   3.17         0.00   3.24     3.30    2.67     3.22   

       Romance  Sci-Fi  Thriller  
12414     4.16    3.58      3.67  
7937      3.81    3.92      3.96  
41483     3.00    1.75      2.50  
29113     2.34    2.77      2.52  
41249     3.16    3.19      3.06

In [7]:
df_training_users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50884 entries, 0 to 50883
Data columns (total 17 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   user id       50884 non-null  int64  
 1   rating count  50884 non-null  int64  
 2   rating ave    50884 non-null  float64
 3   Action        50884 non-null  float64
 4   Adventure     50884 non-null  float64
 5   Animation     50884 non-null  float64
 6   Children      50884 non-null  float64
 7   Comedy        50884 non-null  float64
 8   Crime         50884 non-null  float64
 9   Documentary   50884 non-null  float64
 10  Drama         50884 non-null  float64
 11  Fantasy       50884 non-null  float64
 12  Horror        50884 non-null  float64
 13  Mystery       50884 non-null  float64
 14  Romance       50884 non-null  float64
 15  Sci-Fi        50884 non-null  float64
 16  Thriller      50884 non-null  float64
dtypes: float64(15), int64(2)
memory usage: 6.6 MB


In [8]:
df_training_users['user id'].value_counts()

user id
414    1245
68      967
610     947
249     920
232     861
       ... 
194       1
493       1
3         1
502       1
342       1
Name: count, Length: 397, dtype: int64

**There are 50884 rows to match with the movie dataset and the rating dataset but only 397 viewers**

### Rating Dataset

In [9]:
df_target = pd.read_csv('./data/ratings_for_training.csv', header=None)
df_target.columns = ['ratings']
df_target.sample(5)

ratings
7229       4.0
2425       3.0
44697      4.5
36229      4.0
1409       4.0

*The first row of viewer dataset is the viewer who rated the film from the first row of movie dataset 
a point of ... which is the first row of rating dataset*

In [10]:
df_target.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50884 entries, 0 to 50883
Data columns (total 1 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   ratings  50884 non-null  float64
dtypes: float64(1)
memory usage: 397.7 KB


### Utility Matrix

In [11]:
movie_id = df_training_items['movie id']
user_id = df_training_users['user id']

user_rated_movie = pd.concat([movie_id, user_id, df_target], axis=1)
user_rated_movie.sample(5)

movie id  user id  ratings
29101      7153      365      3.0
11785    109487      153      0.5
16528     45447      226      4.0
23604      4896      305      3.0
15141     69122      212      3.0

In [12]:
utility_matrix = user_rated_movie.pivot_table(index='movie id', columns='user id', values='ratings')
utility_matrix

user id   2    3    4    7    9    10   12   13   15   16   ...  598  599  \
movie id                                                    ...             
4054      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
4069      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
4148      NaN  NaN  NaN  NaN  NaN  NaN  NaN  4.0  NaN  NaN  ...  NaN  2.5   
4149      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  2.5   
4153      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
174055    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  4.0   
176371    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  3.5   
177765    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
179819    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
187593    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   

user id   600  601  603  605  606  607  608  610  
movie id                                          
4054      NaN  NaN  NaN  NaN  3.5  3.0  3.0  2.0  
4069      2.5  NaN  NaN  NaN  2.5  3.0  NaN  NaN  
4148      NaN  NaN  2.0  NaN  2.5  NaN  4.5  3.5  
4149      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4153      NaN  NaN  NaN  NaN  NaN  NaN  NaN  3.0  
...       ...  ...  ...  ...  ...  ...  ...  ...  
174055    NaN  4.0  NaN  NaN  NaN  NaN  NaN  NaN  
176371    NaN  4.0  NaN  NaN  NaN  NaN  NaN  NaN  
177765    NaN  4.5  NaN  NaN  NaN  NaN  NaN  NaN  
179819    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
187593    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[847 rows x 397 columns]

# Data Preprocessing - Scaling

In [13]:
scalerItem = StandardScaler()
items_scaled = scalerItem.fit_transform(df_training_items)
items_scaled = items_scaled[:, 3:]

scalerUser = StandardScaler()
users_scaled = scalerUser.fit_transform(df_training_users)
users_scaled = users_scaled[:, 3:]

scalerTarget = MinMaxScaler(feature_range=(-1, 1))
ratings_scaled = scalerTarget.fit_transform(df_target)

In [14]:
items_scaled.shape, users_scaled.shape, ratings_scaled.shape

((50884, 14), (50884, 14), (50884, 1))

In [15]:
item_train, item_test = train_test_split(items_scaled, train_size=0.80, shuffle=True, random_state=1)
user_train, user_test = train_test_split(users_scaled, train_size=0.80, shuffle=True, random_state=1)
y_train, y_test = train_test_split(ratings_scaled, train_size=0.80, shuffle=True, random_state=1)

item_train.shape, user_train.shape, y_train.shape

((40707, 14), (40707, 14), (40707, 1))

# Training

### Building Hidden Layers

In [16]:
n_outputs = 32

user_NN = keras.Sequential([
    keras.layers.Dense(256, activation='relu'),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(n_outputs)
])

item_NN = keras.Sequential([
    keras.layers.Dense(256, activation='relu'),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(n_outputs)
])

### Building Input and Output Layers

In [17]:
n_item_features = item_train.shape[1]
n_user_features = user_train.shape[1]

input_item = keras.layers.Input(shape=(n_item_features, ))
output_item = item_NN(input_item)

input_user = keras.layers.Input(shape=(n_user_features, ))
output_user = user_NN(input_user)

# Dot Product
final_output = keras.layers.Dot(axes=1, normalize=True)([output_item, output_user])  
# If normalized (vector length = 1), then dot product = cosine similarity (ranges from -1 to 1)

### Complete Model

In [18]:
model = keras.models.Model(inputs=[input_item, input_user], outputs=final_output)
model.summary()

Model: "functional_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                  ┃ Output Shape              ┃         Param # ┃ Connected to               ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)      │ (None, 14)                │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ input_layer_2 (InputLayer)    │ (None, 14)                │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ sequential_1 (Sequential)     │ (None, 32)                │          40,864 │ input_layer[0][0]          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ sequential (Sequential)       │ (None, 32)                │          40,864 │ input_layer_2[0][0]        │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ dot (Dot)                     │ (None, 1)                 │               0 │ sequential_1[0][0],        │
│                               │                           │                 │ sequential[0][0]           │
└───────────────────────────────┴───────────────────────────┴─────────────────┴────────────────────────────┘

 Total params: 81,728 (319.25 KB)

 Trainable params: 81,728 (319.25 KB)

 Non-trainable params: 0 (0.00 B)

### Training

In [19]:
tf.random.set_seed(42)

model.compile(optimizer=keras.optimizers.Adam(learning_rate=0.01),
              loss=keras.losses.MeanSquaredError()
             )

model.fit(x=[item_train, user_train], y=y_train, epochs=30)

Epoch 1/30
1273/1273 ━━━━━━━━━━━━━━━━━━━━ 7s 3ms/step - loss: 0.1438
Epoch 2/30
1273/1273 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - loss: 0.1250
Epoch 3/30
1273/1273 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - loss: 0.1192
Epoch 4/30
1273/1273 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - loss: 0.1149
Epoch 5/30
1273/1273 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - loss: 0.1111
Epoch 6/30
1273/1273 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - loss: 0.1079
Epoch 7/30
1273/1273 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - loss: 0.1056
Epoch 8/30
1273/1273 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - loss: 0.1037
Epoch 9/30
1273/1273 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - loss: 0.1020
Epoch 10/30
1273/1273 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - loss: 0.1005
Epoch 11/30
1273/1273 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 0.0991
Epoch 12/30
1273/1273 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - loss: 0.0977
Epoch 13/30
1273/1273 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - loss: 0.0965
Epoch 14/30
1273/1273 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 0.0954
Epoch 15/30
1273/1273 ━━━━━━━

In [20]:
model.evaluate([item_test, user_test], y_test)

319/319 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.1010


0.09844708442687988

# Recommend Movies to a New Viewer

### Movie Info Dataset

In [21]:
df_distinct_items = df_training_items.drop_duplicates(keep='first', ignore_index=True)
df_distinct_items.shape, df_distinct_items['movie id'].nunique()

((847, 17), 847)

In [22]:
df_item_info = pd.read_csv('./data/movie_info.csv')
df_item_info.head()

movieId                                 title                  genres
0     4054            Save the Last Dance (2001)           Drama|Romance
1     4069           Wedding Planner, The (2001)          Comedy|Romance
2     4148                       Hannibal (2001)         Horror|Thriller
3     4149  Saving Silverman (Evil Woman) (2001)          Comedy|Romance
4     4153                  Down to Earth (2001)  Comedy|Fantasy|Romance

In [23]:
df_item_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 847 entries, 0 to 846
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   movieId  847 non-null    int64 
 1   title    847 non-null    object
 2   genres   847 non-null    object
dtypes: int64(1), object(2)
memory usage: 20.0+ KB


In [24]:
df_item_info.movieId.nunique()

847

**847 rows and 847 movies, no duplicates**

### Creating a new user

In [25]:
new_user_id = 5000
new_rating_count = 3
new_rating_ave = 0.0
new_action = 0.0
new_adventure = 5.0
new_animation = 0.0
new_childrens = 0.0
new_comedy = 0.0
new_crime = 0.0
new_documentary = 0.0
new_drama = 0.0
new_fantasy = 5.0
new_horror = 0.0
new_mystery = 0.0
new_romance = 0.0
new_scifi = 0.0
new_thriller = 0.0

new_user = np.array([[new_user_id, new_rating_count, new_rating_ave,
                      new_action, new_adventure, new_animation, new_childrens,
                      new_comedy, new_crime, new_documentary,
                      new_drama, new_fantasy, new_horror, new_mystery,
                      new_romance, new_scifi, new_thriller]])
new_user.shape

(1, 17)

*The new user enjoys movies from the `adventure`, `fantasy` genres. Let's find the top-rated movies for the new user*

In [26]:
# Replicate the new user to match the number of new movies in the dataset.
new_user_replicated = np.tile(new_user, (len(df_distinct_items), 1))  
# Repeat the array 847 times along axis 0 and 1 times along axis 1
new_user_replicated.shape

(847, 17)

### Predicting

In [27]:
new_user_scaled = scalerUser.transform(new_user_replicated)
new_user_scaled = new_user_scaled[:, 3:]

distinct_items_scaled = scalerItem.transform(df_distinct_items)
distinct_items_scaled = distinct_items_scaled[:, 3:]

new_user_scaled.shape, distinct_items_scaled.shape
# warning because new_user_replicated doesn't have feature names

C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


((847, 14), (847, 14))

In [28]:
ratings_predicted = model.predict([distinct_items_scaled, new_user_scaled])
ratings_predicted = scalerTarget.inverse_transform(ratings_predicted)
ratings_predicted.shape  # has the same order as distinct_items_scaled

27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step


(847, 1)

In [29]:
# must be converted into series to concat
ratings_series = pd.Series(ratings_predicted.reshape(-1), name='rating_predicted')
ratings_df = pd.concat([ratings_series, df_distinct_items], axis=1)

In [30]:
df_merged = pd.merge(df_item_info, ratings_df, left_on='movieId', right_on='movie id', how='inner')
df_merged = df_merged[['rating_predicted', 'ave rating', 'movie id', 'title', 'genres']]
df_merged.sort_values(by='rating_predicted', axis=0, ascending=False, inplace=True)
df_merged.head(10)

rating_predicted  ave rating  movie id  \
367          3.956845    3.816901     40815   
160          3.776751    3.900000      6283   
133          3.560920    3.598039      5816   
604          3.560920    2.875000     74789   
254          3.560920    3.913978      8368   
717          3.560920    3.812500     98809   
68           3.560920    4.106061      4993   
749          3.560920    3.580000    106489   
142          3.560920    4.021277      5952   
793          3.560920    3.416667    118696   

                                                                title  \
367                        Harry Potter and the Goblet of Fire (2005)   
160  Cowboy Bebop: The Movie (Cowboy Bebop: Tengoku no Tobira) (2001)   
133                    Harry Potter and the Chamber of Secrets (2002)   
604                                        Alice in Wonderland (2010)   
254                   Harry Potter and the Prisoner of Azkaban (2004)   
717                         Hobbit: An Unexpected Journey, The (2012)   
68          Lord of the Rings: The Fellowship of the Ring, The (2001)   
749                       Hobbit: The Desolation of Smaug, The (2013)   
142                     Lord of the Rings: The Two Towers, The (2002)   
793                  The Hobbit: The Battle of the Five Armies (2014)   

                               genres  
367        Adventure|Fantasy|Thriller  
160  Action|Animation|Sci-Fi|Thriller  
133                 Adventure|Fantasy  
604                 Adventure|Fantasy  
254                 Adventure|Fantasy  
717                 Adventure|Fantasy  
68                  Adventure|Fantasy  
749                 Adventure|Fantasy  
142                 Adventure|Fantasy  
793                 Adventure|Fantasy

In [31]:
df_merged['movie id'].nunique()

847

# Finding Similar Movies

### Computing Output Item Features

In [32]:
n_item_features = 14
# input layer
input_item = tf.keras.layers.Input(shape=(n_item_features,  ))
# use the pre-trained item_NN
output_item = item_NN(input_item)
# normalize the output
output_item = keras.layers.Lambda(lambda x: tf.linalg.l2_normalize(x, axis=1))(output_item)

item_model = keras.models.Model(input_item, output_item)                                
item_model.summary()

Model: "functional_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ input_layer_4 (InputLayer)           │ (None, 14)                  │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ sequential_1 (Sequential)            │ (None, 32)                  │          40,864 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ lambda (Lambda)                      │ (None, 32)                  │               0 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 40,864 (159.62 KB)

 Trainable params: 40,864 (159.62 KB)

 Non-trainable params: 0 (0.00 B)

In [33]:
output_item_vectors = item_model.predict(distinct_items_scaled)
output_item_vectors.shape

27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step


(847, 32)

### Computing Distance between Points / Movies

In [34]:
dot_products = np.dot(output_item_vectors, output_item_vectors.T)
dot_products.shape

(847, 847)

In [35]:
vector_squared_lengths = np.sum(np.square(output_item_vectors), axis=1)
vector_squared_lengths.shape

(847,)

In [36]:
col = vector_squared_lengths[:, np.newaxis]
row = vector_squared_lengths[np.newaxis, :]

col.shape, row.shape

((847, 1), (1, 847))

In [37]:
squared_distance_matrix = col + row - 2 * dot_products
distance_matrix = np.sqrt(np.abs(squared_distance_matrix))
distance_matrix.shape

(847, 847)

In [38]:
distance_matrix = pd.DataFrame(distance_matrix, columns=None).round(2).abs()
distance_matrix.columns = df_distinct_items['movie id']
distance_matrix.index = df_distinct_items['movie id']

distance_matrix

movie id  6874    8798    46970   48516   58559   60756   68157   71535   \
movie id                                                                   
6874        0.00    0.37    0.69    0.70    0.52    0.86    0.49    0.88   
8798        0.37    0.00    0.49    0.60    0.49    0.62    0.37    0.94   
46970       0.69    0.49    0.00    0.86    0.68    0.30    0.61    1.07   
48516       0.70    0.60    0.86    0.00    0.56    0.86    0.51    0.93   
58559       0.52    0.49    0.68    0.56    0.00    0.77    0.34    0.86   
...          ...     ...     ...     ...     ...     ...     ...     ...   
51412       0.67    0.56    0.52    0.89    0.64    0.60    0.69    1.11   
85510       1.00    1.04    0.97    1.26    1.25    1.10    1.23    1.09   
93363       0.80    0.82    0.95    1.06    0.77    1.02    0.89    1.05   
111364      0.80    0.82    0.95    1.06    0.77    1.02    0.89    1.05   
5128        1.48    1.48    1.41    1.67    1.56    1.49    1.54    1.18   

movie id  74458   77455   ...  31221   36708   37380   46335   7150    51412   \
movie id                  ...                                                   
6874        1.02    1.17  ...    0.98    0.94    0.95    0.37    0.86    0.67   
8798        0.98    1.15  ...    0.74    1.02    0.81    0.00    0.62    0.56   
46970       1.08    1.24  ...    0.56    1.08    0.94    0.49    0.30    0.52   
48516       0.75    0.96  ...    0.86    1.04    0.93    0.60    0.86    0.89   
58559       1.03    1.00  ...    0.75    0.76    0.82    0.49    0.77    0.64   
...          ...     ...  ...     ...     ...     ...     ...     ...     ...   
51412       1.00    1.26  ...    0.64    0.87    0.82    0.56    0.60    0.00   
85510       1.08    1.38  ...    1.25    1.35    1.17    1.04    1.10    0.97   
93363       1.16    1.20  ...    0.97    0.84    0.76    0.82    1.02    0.56   
111364      1.16    1.20  ...    0.97    0.84    0.76    0.82    1.02    0.56   
5128        1.64    1.35  ...    1.58    1.61    1.37    1.48    1.49    1.52   

movie id  85510   93363   111364  5128    
movie id                                  
6874        1.00    0.80    0.80    1.48  
8798        1.04    0.82    0.82    1.48  
46970       0.97    0.95    0.95    1.41  
48516       1.26    1.06    1.06    1.67  
58559       1.25    0.77    0.77    1.56  
...          ...     ...     ...     ...  
51412       0.97    0.56    0.56    1.52  
85510       0.00    1.08    1.08    1.15  
93363       1.08    0.00    0.00    1.44  
111364      1.08    0.00    0.00    1.44  
5128        1.15    1.44    1.44    0.00  

[847 rows x 847 columns]

In [39]:
# Create a mask
mask = np.tril(np.ones_like(distance_matrix, dtype=bool), k=0)
mask

array([[ True, False, False, ..., False, False, False],
       [ True,  True, False, ..., False, False, False],
       [ True,  True,  True, ..., False, False, False],
       ...,
       [ True,  True,  True, ...,  True, False, False],
       [ True,  True,  True, ...,  True,  True, False],
       [ True,  True,  True, ...,  True,  True,  True]])

In [40]:
distance_matrix[mask] = np.inf
distance_matrix

movie id  6874    8798    46970   48516   58559   60756   68157   71535   \
movie id                                                                   
6874         inf    0.37    0.69    0.70    0.52    0.86    0.49    0.88   
8798         inf     inf    0.49    0.60    0.49    0.62    0.37    0.94   
46970        inf     inf     inf    0.86    0.68    0.30    0.61    1.07   
48516        inf     inf     inf     inf    0.56    0.86    0.51    0.93   
58559        inf     inf     inf     inf     inf    0.77    0.34    0.86   
...          ...     ...     ...     ...     ...     ...     ...     ...   
51412        inf     inf     inf     inf     inf     inf     inf     inf   
85510        inf     inf     inf     inf     inf     inf     inf     inf   
93363        inf     inf     inf     inf     inf     inf     inf     inf   
111364       inf     inf     inf     inf     inf     inf     inf     inf   
5128         inf     inf     inf     inf     inf     inf     inf     inf   

movie id  74458   77455   ...  31221   36708   37380   46335   7150    51412   \
movie id                  ...                                                   
6874        1.02    1.17  ...    0.98    0.94    0.95    0.37    0.86    0.67   
8798        0.98    1.15  ...    0.74    1.02    0.81    0.00    0.62    0.56   
46970       1.08    1.24  ...    0.56    1.08    0.94    0.49    0.30    0.52   
48516       0.75    0.96  ...    0.86    1.04    0.93    0.60    0.86    0.89   
58559       1.03    1.00  ...    0.75    0.76    0.82    0.49    0.77    0.64   
...          ...     ...  ...     ...     ...     ...     ...     ...     ...   
51412        inf     inf  ...     inf     inf     inf     inf     inf     inf   
85510        inf     inf  ...     inf     inf     inf     inf     inf     inf   
93363        inf     inf  ...     inf     inf     inf     inf     inf     inf   
111364       inf     inf  ...     inf     inf     inf     inf     inf     inf   
5128         inf     inf  ...     inf     inf     inf     inf     inf     inf   

movie id  85510   93363   111364  5128    
movie id                                  
6874        1.00    0.80    0.80    1.48  
8798        1.04    0.82    0.82    1.48  
46970       0.97    0.95    0.95    1.41  
48516       1.26    1.06    1.06    1.67  
58559       1.25    0.77    0.77    1.56  
...          ...     ...     ...     ...  
51412       0.97    0.56    0.56    1.52  
85510        inf    1.08    1.08    1.15  
93363        inf     inf    0.00    1.44  
111364       inf     inf     inf    1.44  
5128         inf     inf     inf     inf  

[847 rows x 847 columns]

### Finding Similar Movies

In [41]:
sorted_distances = distance_matrix.iloc[10].sort_values(ascending=True)
sorted_distances

movie id
6003     0.41
57669    0.41
48780    0.44
4878     0.44
68237    0.44
         ... 
58559     inf
48516     inf
46970     inf
8798      inf
6874      inf
Name: 79132, Length: 847, dtype: float32

In [42]:
sorted_distances.head(5).index  

Index([6003, 57669, 48780, 4878, 68237], dtype='int64', name='movie id')

In [43]:
n_movies = 20
n_similar_movies = 5

similar_item_ids = []
for i in range(n_movies):
    main_id = distance_matrix.index[i]
    
    sorted_distances = distance_matrix.iloc[i].sort_values(ascending=True)
    similar_ids = sorted_distances.head(n_similar_movies).index
    
    similar_item_ids.append((main_id, similar_ids.tolist()))

similar_item_ids

[(6874, [58295, 8665, 36519, 54286, 7439]),
 (8798, [46335, 32029, 36529, 37733, 4901]),
 (46970, [68793, 4701, 5621, 119141, 4161]),
 (48516, [55765, 84392, 5329, 6552, 4776]),
 (58559, [54997, 4344, 65514, 99114, 4448]),
 (60756, [4974, 5481, 45666, 86911, 107348]),
 (68157, [4448, 4958, 40278, 99114, 115210]),
 (71535, [119145, 8370, 59784, 51255, 52281]),
 (74458, [8783, 8950, 51086, 104879, 45447]),
 (77455, [93840, 52281, 61240, 44972, 6888]),
 (79132, [6003, 57669, 48780, 4878, 68237]),
 (80489, [88129, 84392, 115569, 55765, 5108]),
 (80906, [34072, 5669, 54881, 6331, 7156]),
 (89774, [5810, 27815, 47099, 112290, 4903]),
 (91529, [136020, 6157, 62849, 33794, 6378]),
 (91658, [40583, 6281, 46723, 5400, 112556]),
 (99114, [104913, 4958, 65514, 115210, 4448]),
 (106782, [6708, 5991, 5265, 61323, 129354]),
 (109487, [164179, 176371, 112623, 135143, 69951]),
 (112552, [166643, 96829, 112290, 63876, 106100])]

In [44]:
info = df_item_info[df_item_info['movieId'] == 119145]
info

movieId                                title  \
795   119145  Kingsman: The Secret Service (2015)   

                            genres  
795  Action|Adventure|Comedy|Crime

In [45]:
info['title'].values

array(['Kingsman: The Secret Service (2015)'], dtype=object)

In [46]:
rows = []

for main_id, similar_ids in similar_item_ids:
    
    main_info = df_item_info[df_item_info['movieId'] == main_id]  # show values in DF format
    title = main_info['title'].values[0] if not main_info.empty else None  # take value in column 'title'
    genres = main_info['genres'].values[0] if not main_info.empty else None

    row = {
        #'movie_ID': main_id,
        'title': title,
        'genres': genres
    }

    for loop, id in enumerate(similar_ids, start=1):
        similar_info = df_item_info[df_item_info['movieId'] == id]
        similar_title = similar_info['title'].values[0] if not similar_info.empty else None
        similar_genres = similar_info['genres'].values[0] if not similar_info.empty else None
        
        #row[f'similar_ID_{loop}'] = id
        row[f'similar_title_{loop}'] = similar_title
        row[f'similar_genres_{loop}'] = similar_genres

    rows.append(row)

In [47]:
df_similar_items = pd.DataFrame(rows)
df_similar_items.head(20)

title  \
0                             Kill Bill: Vol. 1 (2003)   
1                                    Collateral (2004)   
2   Talladega Nights: The Ballad of Ricky Bobby (2006)   
3                                 Departed, The (2006)   
4                              Dark Knight, The (2008)   
5                                 Step Brothers (2008)   
6                          Inglourious Basterds (2009)   
7                                    Zombieland (2009)   
8                                Shutter Island (2010)   
9                    Exit Through the Gift Shop (2010)   
10                                    Inception (2010)   
11                                    Town, The (2010)   
12                                   Inside Job (2010)   
13                                      Warrior (2011)   
14                       Dark Knight Rises, The (2012)   
15             Girl with the Dragon Tattoo, The (2011)   
16                             Django Unchained (2012)   
17                     Wolf of Wall Street, The (2013)   
18                                 Interstellar (2014)   
19                                     Whiplash (2014)   

                                        genres  \
0                        Action|Crime|Thriller   
1                  Action|Crime|Drama|Thriller   
2                                Action|Comedy   
3                         Crime|Drama|Thriller   
4                           Action|Crime|Drama   
5                                       Comedy   
6                                 Action|Drama   
7                         Action|Comedy|Horror   
8                       Drama|Mystery|Thriller   
9                           Comedy|Documentary   
10  Action|Crime|Drama|Mystery|Sci-Fi|Thriller   
11                        Crime|Drama|Thriller   
12                                 Documentary   
13                                       Drama   
14                      Action|Adventure|Crime   
15                              Drama|Thriller   
16                                Action|Drama   
17                          Comedy|Crime|Drama   
18                                      Sci-Fi   
19                                       Drama   

                                                                similar_title_1  \
0                                                          Bank Job, The (2008)   
1   Fast and the Furious: Tokyo Drift, The (Fast and the Furious 3, The) (2006)   
2                         Night at the Museum: Battle of the Smithsonian (2009)   
3                                                      American Gangster (2007)   
4                                                           3:10 to Yuma (2007)   
5                                                 Not Another Teen Movie (2001)   
6                                                             Score, The (2001)   
7                                           Kingsman: The Secret Service (2015)   
8                                                           Village, The (2004)   
9                                                Cabin in the Woods, The (2012)   
10                                       Confessions of a Dangerous Mind (2002)   
11                                                                 Drive (2011)   
12                      March of the Penguins (Marche de l'empereur, La) (2005)   
13                                                                8 Mile (2002)   
14                                                               Spectre (2015)   
15                                                               Syriana (2005)   
16                                                                  Rush (2013)   
17                                                        Matchstick Men (2003)   
18                                                               Arrival (2016)   
19                                                        Hidden Figures (2016)   

                 similar_genres_1  